Outlet inflow prevention implementation

rubenzorrilla opened the pull request:
An extra term has been added to the RHS of the navier_stokes_wall_condition to prevent inflow in outlet BCs.

rubenzorrilla committed:
Outlet inflow prevention 

Modifications in the Navier-Stokes condition to avoid unstabilities when outlet inflow occurs.

rubenzorrilla committed:
Creaction of CHARACTERISTIC_VELOCITY var

Creation of a new CHARACTERISTIC_VELOCITY variable. This variable is intended to store the characteristic velocity of the model that has to be used in the computation of the outlet inflow correction.

rubenzorrilla committed:
Advances in the outlet inflow correction computation

Now the characteristic velocity is taken from the ProcessInfo. Besides the inflow correction is multiplied by rho to have consistency.

rubenzorrilla committed:
Inlet process update and rename

Now the inlet process uses the "experimental" processes. Besides the INLET flag is set in both the nodes and conditions of the inlet model part.

rubenzorrilla committed:
Slip conditions process update

Now the apply slip BC process sets the flag SLIP in both the nodes and conditions of the slip model part.

rubenzorrilla committed:
Apply outlet process update

Now the outlet process uses the "experimental" processes in the core for setting both the PRESSURE (set and fixed) and the EXTERNAL_PRESSURE (just set) variables. Besides, the flag OUTLET is set in both the outlet model part nodes and conditions.  
Finally, the average outlet velocity is computed and stored in the CHARACTERISTIC_VELOCITY variable (this value is intended to be u sed in the outlet inflow prevention condition recently implemented).

rubenzorrilla committed:
Merged feature-fldyn-old-processes-update into feature-fldyn-outflow-bc

jcotela commented on applications/FluidDynamicsApplication/custom_conditions/navier_stokes_wall_condition.cpp:
```diff
@@ -139,21 +139,51 @@ void NavierStokesWallCondition<TDim,TNumNodes>::ComputeGaussPointRHSContribution
+    // Gauss pt. outflow condition contribution
+    const double delta = data.delta;
+    const double U_0 = data.charVel;
+    const double S_0 = 0.5*(1-tanh(vGaussProj/U_0/delta));
```
vGaussProj/(U_0*delta) should be cheaper

rubenzorrilla replied:
Done!

jcotela commented on applications/FluidDynamicsApplication/custom_conditions/navier_stokes_wall_condition.cpp:
```diff
@@ -139,21 +139,51 @@ void NavierStokesWallCondition<TDim,TNumNodes>::ComputeGaussPointRHSContribution
+
+    for (unsigned int i=0; i<TNumNodes; ++i)
+    {
+        if (rGeom[i].Is(OUTLET))
```
I think it would be better to calculate this term based on the condition, not the node, since it represents a boundary integral: You want it to act over all triangles on the outlet.
Therefore my recommendation would be to perform the operations in this function only when the CONDITITON is flagged as OUTLET. Such condition will then always (?) act on all its nodes.

rubenzorrilla replied:
Totally agree. Yes, I think that the condition should act on all its nodes (even in the corner and edge nodes).

jcotela reviewed (changes requested):
The condition code should only be executed when the condition is marked as OUTLET (please also make sure that you are marking the condition, not only the nodes)

rubenzorrilla committed:
Outlet inflow prevention through conditions

Now the outlet inflow prevention term is computed according to the conditions OUTLET flag (instead of the nodal one).

rubenzorrilla committed:
RHS contributions in separated functions

The RHS neumann BC contributions addition and the outlet inflow prevention contributions are now computed in two separated functions.

rubenzorrilla committed:
RHS contributions in separated functions

The RHS neumann BC contributions addition and the outlet inflow prevention contributions are now computed in two separated functions.

rubenzorrilla committed:
Merged branch feature-fldyn-outflow-bc into feature-fldyn-outflow-bc
